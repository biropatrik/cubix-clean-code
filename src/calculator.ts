export class Calculator {

    public Add(a: number, b: number) {
        return a + b;
    }

}