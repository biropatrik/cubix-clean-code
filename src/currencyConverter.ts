import { AmountValidationError } from "./exceptions/AmountValidationError";
import { ExchangeRateValidationError } from "./exceptions/ExchangeRateValidationError";
import { UnknownError } from "./exceptions/UnknownError";
import { IExchangeRateService } from "./exchangeRateService";

export class CurrencyConverter {

    private readonly FIXED_AMOUNT = 100;

    constructor(private exchangeRateService: IExchangeRateService) { }

    public Convert(amount: number, fromCurrency: string, toCurrency: string): number {
        const exchangeRate = this.getExchangeRate(fromCurrency, toCurrency);

        try {
            this.validateAmount(amount);
        } catch (error) {
            if (error instanceof Error) {
                throw new AmountValidationError(error.message);
            }
            throw new UnknownError('Unknown error happened');
        }

        try {
            this.validateExchangeRate(exchangeRate);
        } catch (error) {
            if (error instanceof Error) {
                throw new ExchangeRateValidationError(error.message);
            }
            throw new UnknownError('Unknown error happened');
        }

        return amount * exchangeRate;
    }

    public GenerateConversionReport(
        fromCurrency: string,
        toCurrency: string,
        startDate: Date,
        endDate: Date
    ): string {
        const conversions: number[] = [];

        const currentDate = new Date(startDate);

        while (currentDate <= endDate) {
            const exchangeRate = this.exchangeRateService.getExchangeRate(fromCurrency, toCurrency);

            try {
                this.validateExchangeRate(exchangeRate);
            } catch (error) {
                if (error instanceof Error) {
                    throw new ExchangeRateValidationError(error.message);
                }
                throw new UnknownError('Unknown error happened');
            }

            this.calculateConversion(exchangeRate, conversions, currentDate);
        }

        return `Conversion Report:\n${conversions.join('\n')}`;
    }

    private getExchangeRate(fromCurrency: string, toCurrency: string) {
        return this.exchangeRateService.getExchangeRate(fromCurrency, toCurrency);
    }

    private calculateConversion(exchangeRate: number, conversions: number[], currentDate: Date) {
        const convertedAmount = this.FIXED_AMOUNT * exchangeRate; // Assume a fixed amount for simplicity
        conversions.push(convertedAmount);
        currentDate.setDate(currentDate.getDate() + 1);
    }

    private validateExchangeRate(exchangeRate: number) {
        if (!exchangeRate) {
            throw new Error('Unable to fetch exchange rate.');
        }

        if (isNaN(exchangeRate)) {
            throw new Error('Invalid exchange rate.');
        }
    }

    private validateAmount(amount: number) {
        if (isNaN(amount)) {
            throw new Error('Invalid amount input.');
        }
    }
}