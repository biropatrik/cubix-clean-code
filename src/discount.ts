
const STANDARD_DISCOUNT: number = 5;
const SILVER_DISCOUNT: number = 10;
const GOLD_DISCOUNT: number = 15;
const INCREASE_DISCOUNT: number = 5;

abstract class Subscription {
    abstract getDiscount(): number;
}

class Standard extends Subscription {
    getDiscount(): number {
        return STANDARD_DISCOUNT;
    }
}

class Silver extends Subscription {
    getDiscount(): number {
        return SILVER_DISCOUNT;
    }
}

class Gold extends Subscription {
    getDiscount(): number {
        return GOLD_DISCOUNT + this.increaseDiscount();
    }

    private increaseDiscount() {
        return INCREASE_DISCOUNT;
    }
}

export class Discount {

    readonly subscriptions: { [level: string]: new () => Subscription } = {
        'standard': Standard,
        'silver': Silver,
        'gold': Gold
    }

    private createSubscription(level: string): Subscription | undefined {
        const Subscription = this.subscriptions[level];
        return Subscription ? new Subscription() : undefined;
    }

    public calculateDiscountPercentage(level: string): number {
        const subscription = this.createSubscription(level);
        if (subscription === undefined) {
            return 0;
        }
        return subscription.getDiscount();
    }
}