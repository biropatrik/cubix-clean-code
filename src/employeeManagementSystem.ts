class PayrollService {
    readonly employeeService = new EmployeeService();
    readonly salaryService = new SalaryService();

    public calculatePayroll(): number {
        // Simulation to calculate payroll for all employees
        let totalPayroll = 0;
        
        for (const employee of this.employeeService.getAllEmployees()) {
            totalPayroll += this.salaryService.calculateSalary(employee);
        }
        return totalPayroll;
    }
}

class ReportService {

    public generateReports(): string {
        // Simulation to generate various reports
        return "Employee reports: ...";
    }

}

class EmployeeService {
    private employees: Employee[] = [];

    constructor() { }

    public addEmployee(employee: Employee): void {
        // Real-world code to add employee to the system
        this.employees.push(employee);
    }

    public getAllEmployees(): Employee[] {
        return this.employees;
    }

    public promoteEmployee(employee: Employee): void {
        // Code to handle employee promotion logic
    }
}

class SalaryService {
    calculateSalary(employee: Employee): number {
        const newSalary = employee.getSalary() + 1;
        employee.setSalary(newSalary);
        return newSalary;
    }
}
 
class Employee {
    private name: string;
    private salary: number;

    constructor(name: string, salary: number) {
        this.name = name;
        this.salary = salary;
    }

    public getSalary(): number {
        return this.salary;
    }

    public setSalary(salary: number): void {
        this.salary = salary;
    }
}
