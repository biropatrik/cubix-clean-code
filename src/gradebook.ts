export class Gradebook {

    readonly DEFAULT_GRADE = 'D';
    readonly GRADES = new Map<number, string>([
        [70, 'C'],
        [80, 'B'],
        [90, 'A']
    ])

    public calculateGrade(score: number): string {
        return this.getGradeFromMapByScore(this.GRADES, score);
    }

    private getGradeFromMapByScore(grades: Map<number, string>, score: number) {
        let result: string = this.DEFAULT_GRADE;

        for (let key of grades.keys()) {
            result = this.isAGreaterThenEqualB(score, key) ? grades.get(key)! : result;
        }

        return result;
    }

    private isAGreaterThenEqualB(a: number, b: number) {
        return a >= b;
    }
}