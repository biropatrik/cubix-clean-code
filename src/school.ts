class School {
    private classes: SchoolClass[] = [];

    constructor() { }

    addClass(schoolClass: SchoolClass): void {
        this.classes.push(schoolClass);
    }

    getStudentCount(): number {
        let totalStudents = 0;

        for (const schoolClass of this.classes) {
            totalStudents += schoolClass.getStudentCount();
        }
        return totalStudents;
    }
}

class SchoolClass {
    private students: Student[] = [];

    constructor() { }

    addStudent(student: Student): void {
        this.students.push(student);
    }

    getStudents(): Student[] {
        return this.students;
    }

    getStudentCount(): number {
        return this.students.length;
    }
}

class Student {
    private name: string;

    constructor(name: string) {
        this.name = name;
    }

    public getName() {
        return this.name;
    }

    public setName(name: string) {
        this.name = name;
    }
}