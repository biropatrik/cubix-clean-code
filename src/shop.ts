class ShoppingCart {
    private items: CartItem[] = [];

    constructor() { }

    addItem(product: string, price: number): void {
        const newItem = new CartItem(product, price);
        this.items.push(newItem);
    }

    calculateTotal(): number {
        let total = 0;
        for (const item of this.items) {
            total += item.getPrice();
        }
        return total;
    }
}

class CartItem {
    private product: string;
    private price: number;

    constructor(product: string, price: number) {
        this.product = product;
        this.price = price;
    }

    public getProduct() {
        return this.product;
    }

    public getPrice() {
        return this.price;
    }

    public setProduct(product: string) {
        this.product = product;
    }

    public setPrice(price: number) {
        this.price = price;
    }
}