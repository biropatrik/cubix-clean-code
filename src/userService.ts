export class UserService {

    readonly MIN_LENGTH_OF_INPUT = 5;
    readonly MAX_LENGTH_OF_INPUT = 20;
    readonly REGEXP_FOR_INPUT = /^[a-zA-Z0-9]+$/;

    public validateUserInput(input: string): boolean {
        return this.isInputContainsOnlyWhitespaces(input) &&
            this.isInputLenghtBetweenAAndB(input, this.MIN_LENGTH_OF_INPUT, this.MAX_LENGTH_OF_INPUT) &&
            this.validateInputByRegExp(input, this.REGEXP_FOR_INPUT);
    }

    private isInputContainsOnlyWhitespaces(input: string) {
        return input.trim() !== '';
    }

    private isInputLenghtBetweenAAndB(input: string, a: number, b: number) {
        return input.length >= a && input.length <= b;
    }

    private validateInputByRegExp(input: string, regExp: RegExp) {
        return regExp.test(input);
    }

    public processUserData(a: boolean, b: boolean, users: string[], user: string, counter: number): string {
        let result = a && b ? this.findUserInArray(user, users)! : 'No action taken.';
        return !a && b ? this.processCounts(counter) : result;
    }

    private processCounts(counter: number) {
        let result = '';
        for (let i = 0; i < counter; i++) {
            result += 'Processing... ';
        }
        return result;
    }

    private findUserInArray(user: string, users: string[]) {
        for (let i = 0; i < users.length; i++) {
            if (users[i] === user) {
                return 'User found: ' + user + ' at index ' + i;
            }
        }
        return '';
    }

}