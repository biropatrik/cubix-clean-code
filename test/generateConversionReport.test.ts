import { CurrencyConverter } from "../src/currencyConverter";
import { IExchangeRateService } from "../src/exchangeRateService";
import { mock, mockReset } from 'jest-mock-extended';

const mockedExchangeRateService = mock<IExchangeRateService>();

describe('Generate conversion report tests', () => {
    let currencyConverter: CurrencyConverter;

    beforeEach(() => {
        mockReset(mockedExchangeRateService);
        currencyConverter = new CurrencyConverter(mockedExchangeRateService);
    })

    it('should generate conversion report', () => {
        //Arrange
        const numberOfDays = 3;
        let startDate = new Date();
        let endDate = new Date();
        endDate.setDate(startDate.getDate() + (numberOfDays - 1));
        mockedExchangeRateService.getExchangeRate.mockReturnValueOnce(370).mockReturnValueOnce(380).mockReturnValueOnce(400);

        //Act
        const result = currencyConverter.GenerateConversionReport('eur', 'huf', startDate, endDate);

        //Assert
        expect(mockedExchangeRateService.getExchangeRate).toHaveBeenCalledTimes(numberOfDays);
        expect(mockedExchangeRateService.getExchangeRate).toHaveBeenNthCalledWith(1, 'eur', 'huf');
        expect(mockedExchangeRateService.getExchangeRate).toHaveNthReturnedWith(1, 370);

        expect(mockedExchangeRateService.getExchangeRate).toHaveBeenNthCalledWith(2, 'eur', 'huf');
        expect(mockedExchangeRateService.getExchangeRate).toHaveNthReturnedWith(2, 380);

        expect(mockedExchangeRateService.getExchangeRate).toHaveBeenNthCalledWith(3, 'eur', 'huf');
        expect(mockedExchangeRateService.getExchangeRate).toHaveNthReturnedWith(3, 400);

        expect(result).toMatchSnapshot();
    })
})