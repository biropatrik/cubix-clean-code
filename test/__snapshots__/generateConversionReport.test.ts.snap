// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate conversion report tests should generate conversion report 1`] = `
"Conversion Report:
37000
38000
40000"
`;
